libcomposefs/lcfs-mount.c

```diff
@@ -0,0 +1,392 @@
+			LO_NAME_SIZE - 1);
+
+	if (ioctl(loopfd, LOOP_CONFIGURE, &loopconfig) < 0) {
+		errsv = errno;
+		close(loopfd);
+		return -errsv;
+	}
+
+	return loopfd;
+}
+
+static char *compute_lower(const char *imagemount, struct lcfs_mount_state_s *state)
+{
+	size_t size;
+	char *lower;
+	int i;
+
+	/* Compute the total max size (including escapes) */
+	size = 2 * strlen(imagemount);
+	for (i = 0; i < state->options->n_objdirs; i++)
+		size += 1 + 2 * strlen(state->options->objdirs[i]);
+
+	lower = malloc(size + 1);
+	if (lower == NULL)
+		return NULL;
+	*lower = 0;
+
+	escape_mount_option_to(imagemount, lower);
+
+	for (i = 0; i < state->options->n_objdirs; i++) {
+		strcat(lower, ":");
+		escape_mount_option_to(state->options->objdirs[i], lower);
+	}
+
+	return lower;
+}
+
+static int lcfs_mount(struct lcfs_mount_state_s *state)
+{
+	struct lcfs_mount_options_s *options = state->options;
+	struct lcfs_erofs_header_s header;
+	uint32_t image_flags;
+	bool image_has_acls;
+	char imagemountbuf[] = "/tmp/.composefs.XXXXXX";
+	char *imagemount;
+	char loopname[PATH_MAX];
+	int res, errsv;
+	char *lowerdir = NULL;
+	char *upperdir = NULL;
+	char *workdir = NULL;
+	char *overlay_options = NULL;
+	int loopfd;
+	bool require_verity;
+	bool readonly;
+	int mount_flags;
+
+	res = lcfs_validate_verity_fd(state);
+	if (res < 0)
+		return res;
+
+	res = pread(state->fd, &header, sizeof(header), 0);
+	if (res < 0)
+		return -errno;
+
+	if (lcfs_u32_from_file(header.magic) != LCFS_EROFS_MAGIC)
+		return -EINVAL;
+
+	image_flags = lcfs_u32_from_file(header.flags);
+	image_has_acls = (image_flags & LCFS_EROFS_FLAGS_HAS_ACL) != 0;
+
+	require_verity = (options->flags & LCFS_MOUNT_FLAGS_REQUIRE_VERITY) != 0;
+	readonly = (options->flags & LCFS_MOUNT_FLAGS_READONLY) != 0;
+
+	loopfd = setup_loopback(state->fd, state->image_path, loopname);
+	if (loopfd < 0)
+		return loopfd;
+
+	imagemount = mkdtemp(imagemountbuf);
+	if (imagemount == NULL) {
+		errsv = errno;
+		close(loopfd);
+		return -errsv;
+	}
+
+	res = mount(loopname, imagemount, "erofs", MS_RDONLY,
+		    image_has_acls ? "" : "noacl");
+	errsv = errno;
+	close(loopfd);
+	if (res < 0) {
+		rmdir(imagemount);
+		return -errsv;
+	}
+
+	lowerdir = compute_lower(imagemount, state);
+	if (lowerdir == NULL) {
+		res = -ENOMEM;
+		goto fail;
+	}
+
+	if (options->upperdir)
+		upperdir = escape_mount_option(options->upperdir);
+	if (options->workdir)
+		workdir = escape_mount_option(options->workdir);
+
+	res = asprintf(&overlay_options,
+		       "metacopy=on,redirect_dir=on,lowerdir=%s%s%s%s%s%s", lowerdir,
+		       upperdir ? ",upperdir=" : "", upperdir ? upperdir : "",
+		       workdir ? ",workdir=" : "", workdir ? workdir : "",
+		       require_verity ? ",verity=require" : "");
+	if (res < 0) {
+		res = -ENOMEM;
+		goto fail;
+	}
+
+	mount_flags = 0;
+	if (readonly)
+		mount_flags |= MS_RDONLY;
+
+	res = mount("overlay", state->mountpoint, "overlay", mount_flags,
+		    overlay_options);
+
+fail:
+	free(lowerdir);
```

Comment:
Any reason not to cargo cult in the `__attribute__(cleanup)` helpers into this project?

Reply:
I've considered it, and it would be nice. But never got started on it.